template/src/components/Draggable.js

```diff
@@ -0,0 +1,261 @@
+  View,
+  Text,
+  Image,
+  PanResponder,
+  Animated,
+  Dimensions,
+  TouchableOpacity,
+  StyleSheet,
+} from 'react-native'
+
+const clamp = (number, min, max) => Math.max(min, Math.min(number, max))
+
+export default function Draggable(props) {
+  const {
+    renderText = '＋',
+    isCircle,
+    renderSize = 36,
+    imageSource,
+    renderColor,
+    children,
+    shouldReverse = false,
+    disabled = false,
+    debug = false,
+    animatedViewProps,
+    touchableOpacityProps,
+    onDrag = () => {},
+    onShortPressRelease = () => {},
+    onDragRelease = () => {},
+    onLongPress = () => {},
+    onPressIn = () => {},
+    onPressOut = () => {},
+    onRelease = () => {},
+    x = 0,
+    y = 0,
+    z = 1,
+    minX,
+    minY,
+    maxX,
+    maxY,
+  } = props
+
+  // The Animated object housing our xy value so that we can spring back
+  const pan = React.useRef(new Animated.ValueXY())
+  // Always set to xy value of pan, would like to remove
+  const offsetFromStart = React.useRef({x: 0, y: 0})
+  // Width/Height of Draggable (renderSize is arbitrary if children are passed in)
+  const childSize = React.useRef({x: renderSize, y: renderSize})
+  // Top/Left/Right/Bottom location on screen from start of most recent touch
+  const startBounds = React.useRef({top: 0, bottom: 0, left: 0, right: 0})
+  // Whether we're currently dragging or not
+  const isDragging = React.useRef(false)
+
+  const getBounds = React.useCallback(() => {
+    const left = x + offsetFromStart.current.x
+    const top = y + offsetFromStart.current.y
+    return {
+      left,
+      top,
+      right: left + childSize.current.x,
+      bottom: top + childSize.current.y,
+    }
+  }, [x, y])
+
+  const shouldStartDrag = React.useCallback(
+    gs => !disabled && (Math.abs(gs.dx) > 2 || Math.abs(gs.dy) > 2),
+    [disabled],
+  )
+
+  const reversePosition = React.useCallback(() => {
+    Animated.spring(pan.current, {
+      toValue: {x: 0, y: 0},
+      useNativeDriver: false,
+    }).start()
+  }, [pan])
+
+  const onPanResponderRelease = React.useCallback(
+    (e, gestureState) => {
+      isDragging.current = false
+      if (onDragRelease) {
+        onDragRelease(e, gestureState)
+        onRelease(e, true)
+      }
+      if (!shouldReverse) {
+        pan.current.flattenOffset()
+      } else {
+        reversePosition()
+      }
+    },
+    [onDragRelease, shouldReverse, onRelease, reversePosition],
+  )
+
+  const onPanResponderGrant = React.useCallback(
+    (e, gestureState) => {
+      startBounds.current = getBounds()
+      isDragging.current = true
+      if (!shouldReverse) {
+        pan.current.setOffset(offsetFromStart.current)
+        pan.current.setValue({x: 0, y: 0})
+      }
+    },
+    [getBounds, shouldReverse],
+  )
+
+  const handleOnDrag = React.useCallback(
+    (e, gestureState) => {
+      const {dx, dy} = gestureState
+      const {top, right, left, bottom} = startBounds.current
+      const far = 999999999
```

Comment:
what's this?

Reply:
Hi @ThinhKimVo. Could you please answer all questions regarding this component?

Reply:
it represents for **min** and **max** finite in case the min and max prop `isFinite`
```
Number.isFinite(minX) ? minX - left : -far,
Number.isFinite(maxX) ? maxX - right : far,
```